### PR TITLE
i55: Add validation to email sign in and sign up

### LINKED ISSUE #55: Add validation to email sign in and sign up
- use form validation 
- show error for text fields on invalid input 

### sign in with email 
- [x] validate `email` text field on submit and then on each key press (use standard 'email' validator) 
- [x] validate `password` text field on submit and then on each key press -> min = 6, max = 30 
- [x] clear form on tap `create account` chip

### create account 
- [x] validate `email` text field on submit and then on each key press (use standard 'email' validator) 
- [x] validate `password` text field on submit and then on each key press -> min = 6, max = 30, same as `repeat password` field 
- [x] validate `repeat password` text field on submit and then on each key press -> min = 6, max = 30, same as `password` field 
- [x] clear form on tap `sign in` chip

### autovalidation
- [x] add Autovalidate enum into emailAuthPage state, set disabled as default
- [x] on tap sign in or create acount buttons, set autovalidate to autovalidate.onUserInput 
- [x] when change from sign in to sign up, or vice versa, reset autovalidate to autovalidate.disabled

### PR DESCRIPTION
**Pull request by tamari-gray**
fixes #55

### CODE REVIEW

**Comment by tamari-gray** [comment]
2 tests in other_auth_options_page_widget_test.dart are failing due to overflow errors that will be fixed in future issue

File: lib/widgets/auth/other_auth_options_page.dart [hunk 1]
```diff
@@ -37,33 +38,36 @@ class OtherAuthOptionsPage extends StatelessWidget {
...
-                        SizedBox(height: 20),
-                        PasswordTextField(
-                          visible: vm.showPassword,
-                        ),
-                        SizedBox(height: 20),
-                        if (vm.mode == EmailAuthMode.signUp)
-                          RepeatPasswordTextField(
+                    child: Form(
+                      child: Column(
+                        crossAxisAlignment: CrossAxisAlignment.center,
```
Comment (tamari-gray) [inline lib/widgets/auth/other_auth_options_page.dart:43]: wrapped textFeilds in a Form widget

**Review by tamari-gray — COMMENTED** [review]
ready for review :)

**Comment by tamari-gray** [comment]
waiting for #178 to complete this pr

**Comment by nickmeinhold** [comment]
Just adding a quick note: #178 is merged now and @tamari-gray is going to merge `dev` into this branch - I've removed the request for review so I'll get a notification when a review is requested. Thanks!

File: test/widget/auth/other_auth_options_page_widget_test.dart [hunk 2]
```diff
@@ -1,14 +1,8 @@
 import 'package:crowdleague/actions/auth/sign_in_with_apple.dart';
```
Comment (tamari-gray) [inline test/widget/auth/other_auth_options_page_widget_test.dart:1]: all removed code was moved into email_auth_options_widget_test.dart

**Review by tamari-gray — COMMENTED** [review]
ready for review :)

**Comment by nickmeinhold** [comment]
Quick note from manual testing: the autovalidation appears to not be working on textfields in both the email sign in and create account.

**Review by nickmeinhold — CHANGES_REQUESTED** [review]
Awesome work mate!  So many beautiful tests!!!  

I'd like the tests to follow our style guide - I've made some comments to indicate where I thought the tests weren't following the style guide, I think most of it I've made separate issues but there may be some things to do on this PR so when you're free tomorrow, could you go through the comments, clarify anything I didn't communicate well enough, make any changes you agree on, discuss any you don't and request another review when everything's been addressed?  Thanks heaps! 🙂

File: lib/utils/form_validation.dart [hunk 3]
```diff
@@ -0,0 +1,11 @@
+final emailPattern =
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply lib/utils/form_validation.dart:1]: yeep!

File: lib/widgets/auth/email_auth_options_page.dart [hunk 4]
```diff
@@ -111,7 +122,16 @@ class PasswordTextField extends StatelessWidget {
   Widget build(BuildContext context) {
     return Padding(
       padding: const EdgeInsets.only(left: 25.0, right: 25.0),
-      child: TextField(
+      child: TextFormField(
+        validator: (value) {
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply lib/widgets/auth/email_auth_options_page.dart:126]: shes vedy clean now :sunglasses:

File: lib/widgets/auth/email_auth_options_page.dart [hunk 5]
```diff
@@ -187,6 +218,7 @@ class EmailSignInChip extends StatelessWidget {
         label: Text('SIGN IN'),
         selected: _selected,
         onSelected: (bool selected) {
+          Form.of(context).reset();
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply lib/widgets/auth/email_auth_options_page.dart:221]: ooh the formfeild values are being saved in redux, and in the 'Form' object. I just feel better clearing both when changing from sign in form to sign up form

File: test/unit/util/form_validation_test.dart [hunk 6]
```diff
@@ -0,0 +1,41 @@
+// Import the test package and Counter class
+import 'package:crowdleague/utils/form_validation.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('form validation', () {
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/util/form_validation_test.dart:6]: yeeeh!!

[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/unit/util/form_validation_test.dart:6]: tests added :)
so glad to knuckle down on that 🏋️

File: lib/widgets/auth/email_auth_options_page.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (tamari-gray) [reply lib/widgets/auth/email_auth_options_page.dart:221]: please comment if that wasnt clear

File: test/widget/auth/email_auth_options_page_widget_test.dart [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+void main() {
+  group('EmailAuthOptionsPage', () {
+    testWidgets('displays without overflowing', (WidgetTester tester) async {
+      // passing test indicates no overflowing as test suite uses a small device screen
+
+      // Setup the app state with expected values
+      final initialAppState = AppState.init();
+
+      // Create the test harness.
+      final store = Store<AppState>(appReducer, initialState: initialAppState);
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/widget/auth/email_auth_options_page_widget_test.dart:25]: woah noice

[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/widget/auth/email_auth_options_page_widget_test.dart:25]: can i leave this for #225?

File: test/widget/auth/email_auth_options_page_widget_test.dart [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+
+import '../../utils/verify_dispatch_middleware.dart';
+
+void main() {
+  group('EmailAuthOptionsPage', () {
+    testWidgets('displays without overflowing', (WidgetTester tester) async {
+      // passing test indicates no overflowing as test suite uses a small device screen
+
+      // Setup the app state with expected values
+      final initialAppState = AppState.init();
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/widget/auth/email_auth_options_page_widget_test.dart:22]: oh sweet thank you! ill leave it for #226

[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/widget/auth/email_auth_options_page_widget_test.dart:22]: these tests gonna be cleany beany

File: test/widget/auth/email_auth_options_page_widget_test.dart [hunk 9]
```diff
@@ -0,0 +1,525 @@
...
+          testMiddleware.received(UpdateEmailAuthOptionsPage(
+            mode: EmailAuthMode.signIn,
+            email: '',
+            password: '',
+            repeatPassword: '',
+          )),
+          true);
+    });
+
+    testWidgets('EmailTextField dispatches UpdateEmailAuthOptionsPage',
```
[the remark this answers is not in the document]
Comment (tamari-gray) [reply test/widget/auth/email_auth_options_page_widget_test.dart:167]: oh alg! Thought i was doing it the right way keeping all related tests in the test group. But really love extracting these into their own files!
Gonna be alot easier to find everything!
sorry to make you branch out this pr into so many other issues, our process will get faster. I swear!! 😄

**Comment by tamari-gray** [comment]
I have not been looking at the style guide! I was using other files as a reference but from now on ill double check the guide before pushing anything up! :)

File: lib/models/auth/vm_email_auth_options_page.dart [hunk 10]
```diff
@@ -0,0 +1,61 @@
+library vm_email_auth_options_page;
```
Comment (tamari-gray) [inline lib/models/auth/vm_email_auth_options_page.dart:1]: this was renaming vm_other_auth_options_page.dart to vm_email_auth_options_page.dart.
vm_other_auth_options_page.dart was deleted and we added the autovalidate member :)

File: lib/reducers/auth/update_other_auth_options_page.dart [hunk 11]
```diff
@@ -9,17 +9,19 @@ class UpdateEmailAuthOptionsPageReducer
...
-                  action.showPassword ?? a.otherAuthOptionsPage.showPassword;
-              a.otherAuthOptionsPage.step =
-                  action.step ?? a.otherAuthOptionsPage.step;
-              a.otherAuthOptionsPage.email =
-                  action.email ?? a.otherAuthOptionsPage.email;
-              a.otherAuthOptionsPage.password =
-                  action.password ?? a.otherAuthOptionsPage.password;
-              a.otherAuthOptionsPage.repeatPassword = action.repeatPassword ??
-                  a.otherAuthOptionsPage.repeatPassword;
+              a.emailAuthOptionsPage.mode =
```
Comment (tamari-gray) [inline lib/reducers/auth/update_other_auth_options_page.dart:12]: just renaming

**Review by tamari-gray — COMMENTED** [review]
ready for next review 😄

File: lib/utils/form_validation.dart [hunk 12]
```diff
@@ -1,3 +1,4 @@
+// Regex pattern from https://stackoverflow.com/questions/59646163/email-validation-not-working-in-flutter-display-message-even-if-email-address-is
```
Comment (nickmeinhold) [inline lib/utils/form_validation.dart:1]: thanks for doing that!

File: lib/widgets/auth/email_auth_options_page.dart [hunk 13]
```diff
@@ -165,7 +165,7 @@ class RepeatPasswordTextField extends StatelessWidget {
       child: TextFormField(
         validator: (value) {
           if (value.isEmpty) {
-            return 'please enter password';
+            return 'please enter password again';
```
Comment (nickmeinhold) [inline lib/widgets/auth/email_auth_options_page.dart:188]: nice pick up!

File: test/widget/auth/email_auth_options_page_widget_test.dart [hunk 14]
```diff
@@ -521,5 +546,63 @@ void main() {
       // check sign in button didnt dispatch SignInWithEmail action
       expect(testMiddleware.received(SignInWithEmail()), false);
     });
+
+    testWidgets(
```
Comment (nickmeinhold) [inline test/widget/auth/email_auth_options_page_widget_test.dart:553]: Awesome work!!!

File: test/widget/auth/email_auth_options_page_widget_test.dart [hunk 15]
```diff
@@ -389,25 +389,39 @@ void main() {
       final createAccountButton = find.byType(CreateAccountButton);
       expect(createAccountButton, findsOneWidget);
 
+      final emptyEmailFieldError = find.text('please enter email');
```
Comment (nickmeinhold) [inline test/widget/auth/email_auth_options_page_widget_test.dart:395]: Neat trick!

File: lib/models/auth/vm_email_auth_options_page.dart (hunk 10 above)
Comment (nickmeinhold) [reply lib/models/auth/vm_email_auth_options_page.dart:1]: thanks for the helpful commenting!

File: test/widget/auth/email_auth_options_page_widget_test.dart [hunk 16]
```diff
@@ -604,5 +607,72 @@ void main() {
       expect(invalidEmailText, findsNothing);
       expect(invalidPasswordText, findsNothing);
     });
+    testWidgets(
+        'After tap create account with invalid details, form autovalidates on user input',
```
Comment (nickmeinhold) [inline test/widget/auth/email_auth_options_page_widget_test.dart:611]: Loving the tests!

File: lib/reducers/auth/update_other_auth_options_page.dart (hunk 11 above)
Comment (nickmeinhold) [reply lib/reducers/auth/update_other_auth_options_page.dart:12]: lovely!

**Review by nickmeinhold — APPROVED** [review]
Heck yeah!!! Thanks for adding some great extra work after completing the PR... huge effort!  Really happy with the result, thanks!

File: lib/utils/form_validation.dart (hunk 12 above)
Comment (tamari-gray) [reply lib/utils/form_validation.dart:1]: allgood!

File: lib/widgets/auth/email_auth_options_page.dart (hunk 13 above)
Comment (tamari-gray) [reply lib/widgets/auth/email_auth_options_page.dart:188]: shank yu 😄

File: test/widget/auth/email_auth_options_page_widget_test.dart (hunk 14 above)
Comment (tamari-gray) [reply test/widget/auth/email_auth_options_page_widget_test.dart:553]: much abliged sensei 🙇

File: test/widget/auth/email_auth_options_page_widget_test.dart (hunk 15 above)
Comment (tamari-gray) [reply test/widget/auth/email_auth_options_page_widget_test.dart:395]: i have my moments 🤹

File: test/widget/auth/email_auth_options_page_widget_test.dart (hunk 16 above)
Comment (tamari-gray) [reply test/widget/auth/email_auth_options_page_widget_test.dart:611]: 🙇‍♂️

**Pull request merged**